pkgs/metrics/src/lib.zig

```diff
@@ -0,0 +1,202 @@
+};
+
+/// Timer struct returned to the application.
+pub const Timer = struct {
+    start_time: i128,
+    histogram: *const anyopaque, // We'll store which histogram to use
+    is_chain: bool,
+
+    /// Stops the timer and records the duration in the histogram.
+    pub fn observe(self: Timer) void {
+        if (isZKVM()) return;
+
+        const end_time = getTimestamp();
+        const duration_ns = end_time - self.start_time;
+
+        // For freestanding targets where we can't measure time, just record 0
+        const duration_seconds = if (duration_ns == 0) 0.0 else @as(f32, @floatFromInt(duration_ns)) / 1_000_000_000.0;
+
+        if (self.is_chain) {
+            metrics.chain_onblock_duration_seconds.observe(duration_seconds);
+        } else {
+            metrics.block_processing_duration_seconds.observe(duration_seconds);
+        }
+    }
+};
+
+/// A wrapper struct that exposes a `start` function to match the existing API.
+pub const Histogram = struct {
+    is_chain: bool,
+
+    pub fn start(self: *const Histogram) Timer {
+        return Timer{
+            .start_time = getTimestamp(),
+            .histogram = undefined, // Not used in this implementation
+            .is_chain = self.is_chain,
+        };
+    }
+};
+
+/// The public variables the application interacts with.
+/// Calling `.start()` on these will start a new timer.
+pub var chain_onblock_duration_seconds: Histogram = Histogram{ .is_chain = true };
+pub var block_processing_duration_seconds: Histogram = Histogram{ .is_chain = false };
+
+/// Initializes the metrics system. Must be called once at startup.
+pub fn init(allocator: std.mem.Allocator) !void {
+    _ = allocator; // Not needed for basic histograms
+    if (g_initialized) return;
+
+    // For ZKVM targets, use no-op metrics
+    if (isZKVM()) {
+        std.log.info("Using no-op metrics for ZKVM target", .{});
+        g_initialized = true;
+        return;
+    }
+
+    metrics = .{
+        .chain_onblock_duration_seconds = Metrics.ChainHistogram.init("chain_onblock_duration_seconds", .{
+            .help = "Time taken to process a block in the chain's onBlock function."
+        }, .{}),
+        .block_processing_duration_seconds = Metrics.BlockProcessingHistogram.init("block_processing_duration_seconds", .{
+            .help = "Time taken to process a block in the state transition function."
+        }, .{}),
+    };
+
+    g_initialized = true;
+}
+
+/// Writes metrics to a writer (for Prometheus endpoint).
+pub fn writeMetrics(writer: anytype) !void {
+    if (!g_initialized) return error.NotInitialized;
+
+    // For ZKVM targets, write no metrics
+    if (isZKVM()) {
+        try writer.writeAll("# Metrics disabled for ZKVM target\n");
+        return;
+    }
+
+    try metrics_lib.write(&metrics, writer);
+}
+
+/// Starts a simple HTTP server to serve metrics on /metrics endpoint.
+pub fn startListener(allocator: std.mem.Allocator, port: u16) !void {
```

Comment:
this needs to move out from this file, to a file: routes.zig in the same directory, and it needs to accept an httpz instance to setup its route on that server.

httpz instance needs to started on the main cli and metrics need to be initialized/instantiated from there

Reply:
right so @bomanaps we would start the server when running `zeam beam`, possibly guarded by a `--metrics` option.